x-pack/legacy/plugins/siem/server/lib/detection_engine/types.ts

```diff
@@ -22,19 +22,23 @@ export interface ThreatParams {
   technique: IMitreAttack[];
 }
 
+export type RuleType = 'query' | 'saved_query' | 'machine_learning';
+
 export interface RuleAlertParams {
+  anomalyThreshold?: number;
   description: string;
   note: string | undefined | null;
   enabled: boolean;
   falsePositives: string[];
   filters: PartialFilter[] | undefined | null;
   from: string;
   immutable: boolean;
-  index: string[];
+  index: string[] | undefined | null;
   interval: string;
   ruleId: string | undefined | null;
   language: string | undefined | null;
   maxSignals: number;
+  mlJobId?: string;
```

Comment:
There is a subtle typescript "trick" we are doing by saying in the others something like this:

```ts
mlJobId: string | undefined | null;
```

or at lest this:
```ts
mlJobId: string | undefined;
```
vs. just using an optional.

If you explicitly state that it must be at least undefined, it forces the type checker to throw errors on any and all functions that require it and are doing destructoring such as:

```ts
foo({ mlJobId  }: { mlJobId }
```

Requires the programmer to address and add that as a touch point:
```ts
foo({ mlJobId  });
```

Whereas if you don't do that, then it will be easier to miss each touch point as you're saying it is optional and someone could forget which functions should have that value and miss it.

Just letting you know why we added it. Also, even though you are saying `mlJobId?: string` really earlier you are stating in here:

```ts
x-pack/legacy/plugins/siem/server/lib/detection_engine/signals/signal_params_schema.ts

```ts
mlJobId: schema.nullable(schema.string()),
```

What that does with the schema is create a `null` mlJobId but not an `undefined` when you do not specify a `mlJobId` when creating it within the saved objects.

So you really do in some cases will have a `mlJobId` as `null` and not just `undefined` if you re-use this type to represent the Saved Object value.

Reply:
It sounds like you may disagree with how I've been approaching these new fields in general. I tried to summarize the idea in the PR description, but I'll elaborate here.

The way that I've been thinking about these new fields, optional seems to be exactly what I want:
* At the HTTP Level, those fields may expressly be sent by a client, or they may not. I don't want to force the client to provide unnecessary values.
* If we're passing that data along a code path that requires those optional fields (i.e. they're no longer optional), we should (and do) narrow those types with runtime checks.

The nullable stuff with Joi was something I copied without really thinking about it; I personally would prefer not to make those fields null.

> you're saying it is optional and someone could forget which functions should have that value and miss it.
I didn't really understand this statement nor the `foo()` example; could you please elaborate?